From e6a18836cadf16b4d2058dbbdd1b97847aff94b8 Mon Sep 17 00:00:00 2001
From: Eduardo Alberti <eduardo.alberti@windriver.com>
Date: Thu, 10 Aug 2023 20:14:04 +0000
Subject: [PATCH] Enable the usage of Device Plugins

In certain situations, processing resources are made available
through Device Plugins,  either because they are structures to
be isolated, or because  they require  some special treatment.
In these cases,  Intel Power Manager is  not able to recognize
device plugins and treat them as CPU resources.

This modification aims to enable the use of device plugins as
processing resources, whenever this indication is convenient.
To do so,  the user must declare,  during the  application of
PowerConfig, which "customDevices" must  be considered during
the admission of the pod.

This modification also exempts the user from using the "CPU"
directive in resource requests whenever an entry of custom
Device is indicated.

Pull request:
https://github.com/intel/kubernetes-power-manager/pull/38

Signed-off-by: Eduardo Alberti <eduardo.alberti@windriver.com>
---
 api/v1/powerconfig_types.go                   |  3 +
 api/v1/powernode_types.go                     |  3 +
 .../bases/power.intel.com_powerconfigs.yaml   |  5 ++
 .../crd/bases/power.intel.com_powernodes.yaml |  5 ++
 controllers/powerconfig_controller.go         | 18 +++-
 controllers/powernode_controller.go           | 11 +++
 controllers/powerpod_controller.go            | 84 +++++++++++++++----
 7 files changed, 110 insertions(+), 19 deletions(-)

diff --git a/api/v1/powerconfig_types.go b/api/v1/powerconfig_types.go
index 905d2d7..c74c3a4 100644
--- a/api/v1/powerconfig_types.go
+++ b/api/v1/powerconfig_types.go
@@ -33,6 +33,9 @@ type PowerConfigSpec struct {
 
 	// The PowerProfiles that will be created by the Operator
 	PowerProfiles []string `json:"powerProfiles,omitempty"`
+
+	// The CustomDevices include alternative devices that represents CPU resources
+	CustomDevices []string `json:"customDevices,omitempty"`
 }
 
 // PowerConfigStatus defines the observed state of PowerConfig
diff --git a/api/v1/powernode_types.go b/api/v1/powernode_types.go
index 7ee0bd6..9d95de8 100644
--- a/api/v1/powernode_types.go
+++ b/api/v1/powernode_types.go
@@ -42,6 +42,9 @@ type PowerNodeSpec struct {
 	// Information about the containers in the cluster utilizing some PowerWorkload
 	PowerContainers []Container `json:"powerContainers,omitempty"`
 
+	// The CustomDevices include alternative devices that represents CPU resources
+	CustomDevices []string `json:"customDevices,omitempty"`
+
 	// The PowerProfiles in the cluster that are currently being used by Pods
 	//ActiveProfiles map[string]bool `json:"activeProfiles,omitempty"`
 
diff --git a/config/crd/bases/power.intel.com_powerconfigs.yaml b/config/crd/bases/power.intel.com_powerconfigs.yaml
index 0d259af..05db64b 100644
--- a/config/crd/bases/power.intel.com_powerconfigs.yaml
+++ b/config/crd/bases/power.intel.com_powerconfigs.yaml
@@ -46,6 +46,11 @@ spec:
                 items:
                   type: string
                 type: array
+              customDevices:
+                description: Custom Devices define other CPU Resources to be considered in Pod's spec
+                items:
+                  type: string
+                type: array
             type: object
           status:
             description: PowerConfigStatus defines the observed state of PowerConfig
diff --git a/config/crd/bases/power.intel.com_powernodes.yaml b/config/crd/bases/power.intel.com_powernodes.yaml
index 3db1edd..ef6165b 100644
--- a/config/crd/bases/power.intel.com_powernodes.yaml
+++ b/config/crd/bases/power.intel.com_powernodes.yaml
@@ -77,6 +77,11 @@ spec:
                 type: string
               unaffectedCores:
                 type: string
+              customDevices:
+                description: Custom Devices define other CPU Resources to be considered in Pod's spec
+                items:
+                  type: string
+                type: array
             type: object
           status:
             description: PowerNodeStatus defines the observed state of PowerNode
diff --git a/controllers/powerconfig_controller.go b/controllers/powerconfig_controller.go
index 6adf96e..a354ffe 100644
--- a/controllers/powerconfig_controller.go
+++ b/controllers/powerconfig_controller.go
@@ -188,6 +188,13 @@ func (r *PowerConfigReconciler) Reconcile(c context.Context, req ctrl.Request) (
 	labelledNodeList := &corev1.NodeList{}
 	listOption := config.Spec.PowerNodeSelector
 
+	// Searching for Custom Devices in PowerConfig
+	CustomDevices := config.Spec.CustomDevices
+	if len(CustomDevices) > 0 {
+		logger.V(5).Info("The behaviour of Power Node Agent will be affected by the following devices.",
+			"Custom Devices", CustomDevices)
+	}
+
 	logger.V(5).Info("confirming desired nodes match the power node selector")
 	err = r.Client.List(context.TODO(), labelledNodeList, client.MatchingLabels(listOption))
 	if err != nil {
@@ -216,7 +223,8 @@ func (r *PowerConfigReconciler) Reconcile(c context.Context, req ctrl.Request) (
 				}
 
 				powerNodeSpec := &powerv1.PowerNodeSpec{
-					NodeName: node.Name,
+					NodeName:      node.Name,
+					CustomDevices: CustomDevices,
 				}
 
 				powerNode.Spec = *powerNodeSpec
@@ -229,9 +237,17 @@ func (r *PowerConfigReconciler) Reconcile(c context.Context, req ctrl.Request) (
 				return ctrl.Result{}, err
 			}
 		}
+
+		powerNode.Spec.CustomDevices = CustomDevices
+		err := r.Client.Update(context.TODO(), powerNode)
+		if err != nil {
+			logger.Error(err, "Failed to update PowerNode with custom Devices.")
+			return ctrl.Result{}, err
+		}
 	}
 
 	config.Status.Nodes = r.State.PowerNodeList
+	config.Spec.CustomDevices = CustomDevices
 	logger.V(5).Info("configured power node added to the power node list")
 	err = r.Client.Status().Update(context.TODO(), config)
 	if err != nil {
diff --git a/controllers/powernode_controller.go b/controllers/powernode_controller.go
index ee8fba9..9c26a11 100644
--- a/controllers/powernode_controller.go
+++ b/controllers/powernode_controller.go
@@ -78,6 +78,12 @@ func (r *PowerNodeReconciler) Reconcile(c context.Context, req ctrl.Request) (ct
 		}
 		return ctrl.Result{RequeueAfter: queuetime}, err
 	}
+
+	CustomDevices := powerNode.Spec.CustomDevices
+	if len(CustomDevices) > 0 {
+		logger.V(5).Info("The PowerNode contains the following custom devices.", "Custom Devices", powerNode.Spec.CustomDevices)
+	}
+
 	powerProfiles := &powerv1.PowerProfileList{}
 	logger.V(5).Info("retrieving the power profile list")
 	err = r.Client.List(context.TODO(), powerProfiles)
@@ -171,6 +177,11 @@ func (r *PowerNodeReconciler) Reconcile(c context.Context, req ctrl.Request) (ct
 		cores := prettifyCoreList(reservedSystemCpus)
 		powerNode.Spec.UneffectedCores = cores
 	}
+
+	if len(CustomDevices) > 0 {
+		powerNode.Spec.CustomDevices = CustomDevices
+	}
+
 	err = r.Client.Update(context.TODO(), powerNode)
 	if err != nil {
 		return ctrl.Result{RequeueAfter: queuetime}, err
diff --git a/controllers/powerpod_controller.go b/controllers/powerpod_controller.go
index 64334de..c348a42 100644
--- a/controllers/powerpod_controller.go
+++ b/controllers/powerpod_controller.go
@@ -151,11 +151,26 @@ func (r *PowerPodReconciler) Reconcile(c context.Context, req ctrl.Request) (ctr
 		return ctrl.Result{}, podNotRunningErr
 	}
 
+	// Get customDevices that need to be considered in the pod
+	logger.V(5).Info("Retrivieng custom resources from PowerNode")
+	powernode := &powerv1.PowerNode{}
+	err = r.Get(context.TODO(), client.ObjectKey{
+		Namespace: IntelPowerNamespace,
+		Name:      nodeName,
+	}, powernode)
+	if err != nil {
+		if errors.IsNotFound(err) {
+			return ctrl.Result{}, nil
+		}
+		logger.Error(err, "error while trying to retrieve the PowerNode")
+		return ctrl.Result{}, err
+	}
+
 	// Get the containers of the pod that are requesting exclusive CPUs
-	logger.V(5).Info("retrieving the containers requested for the exclusive CPUs")
-	containersRequestingExclusiveCPUs := getContainersRequestingExclusiveCPUs(pod, &logger)
-	if len(containersRequestingExclusiveCPUs) == 0 {
-		logger.Info("no containers are requesting exclusive CPUs")
+	logger.V(5).Info("Retrieving the containers requested for the exclusive CPUs or this/these Custom Devices", "Custom Devices", powernode.Spec.CustomDevices)
+	admissibleContainers := getAdmissibleContainers(pod, &logger, powernode.Spec.CustomDevices)
+	if len(admissibleContainers) == 0 {
+		logger.Info("No containers are requesting exclusive CPUs or Custom Resources")
 		return ctrl.Result{}, nil
 	}
 	podUID := pod.GetUID()
@@ -171,7 +186,7 @@ func (r *PowerPodReconciler) Reconcile(c context.Context, req ctrl.Request) (ctr
 		logger.Error(err, "error retrieving the power profiles from the cluster")
 		return ctrl.Result{}, err
 	}
-	powerProfilesFromContainers, powerContainers, err := r.getPowerProfileRequestsFromContainers(containersRequestingExclusiveCPUs, powerProfileCRs.Items, pod, &logger)
+	powerProfilesFromContainers, powerContainers, err := r.getPowerProfileRequestsFromContainers(admissibleContainers, powerProfileCRs.Items, pod, &logger, powernode.Spec.CustomDevices)
 	logger.V(5).Info("retrieving the power profiles and cores from the pod requests")
 	if err != nil {
 		logger.Error(err, "error retrieving the power profile from the pod requests")
@@ -248,7 +263,7 @@ func (r *PowerPodReconciler) Reconcile(c context.Context, req ctrl.Request) (ctr
 	return ctrl.Result{}, nil
 }
 
-func (r *PowerPodReconciler) getPowerProfileRequestsFromContainers(containers []corev1.Container, profileCRs []powerv1.PowerProfile, pod *corev1.Pod, logger *logr.Logger) (map[string][]uint, []powerv1.Container, error) {
+func (r *PowerPodReconciler) getPowerProfileRequestsFromContainers(containers []corev1.Container, profileCRs []powerv1.PowerProfile, pod *corev1.Pod, logger *logr.Logger, CustomDevices []string) (map[string][]uint, []powerv1.Container, error) {
 
 	logger.V(5).Info("get the power profiles from the containers")
 	_ = context.Background()
@@ -257,7 +272,7 @@ func (r *PowerPodReconciler) getPowerProfileRequestsFromContainers(containers []
 	powerContainers := make([]powerv1.Container, 0)
 	for _, container := range containers {
 		logger.V(5).Info("retrieving the requested power profile from the container spec")
-		profile, err := getContainerProfileFromRequests(container, logger)
+		profile, err := getContainerProfileFromRequests(container, logger, CustomDevices)
 		if err != nil {
 			return map[string][]uint{}, []powerv1.Container{}, err
 		}
@@ -366,7 +381,7 @@ func isContainerInList(name string, containers []powerv1.Container, logger *logr
 	return false
 }
 
-func getContainerProfileFromRequests(container corev1.Container, logger *logr.Logger) (string, error) {
+func getContainerProfileFromRequests(container corev1.Container, logger *logr.Logger, CustomDevices []string) (string, error) {
 	profileName := ""
 	moreThanOneProfileError := errors.NewServiceUnavailable("cannot have more than one power profile per container")
 	resourceRequestsMismatchError := errors.NewServiceUnavailable("mismatch between CPU requests and the power profile requests")
@@ -388,9 +403,31 @@ func getContainerProfileFromRequests(container corev1.Container, logger *logr.Lo
 		powerProfileResourceName := corev1.ResourceName(fmt.Sprintf("%s%s", ResourcePrefix, profileName))
 		numRequestsPowerProfile := container.Resources.Requests[powerProfileResourceName]
 		numLimitsPowerProfile := container.Resources.Limits[powerProfileResourceName]
-		numRequestsCPU := container.Resources.Requests[CPUResource]
-		numLimistCPU := container.Resources.Limits[CPUResource]
-		if numRequestsCPU != numRequestsPowerProfile || numLimistCPU != numLimitsPowerProfile {
+
+		// Selecting resources to search
+		numRequestsCPU := 0
+		numLimitsCPU := 0
+
+		// If the Isolcpu resource is requested, change the CPU request to
+		// allow the check
+		for _, deviceName := range CustomDevices {
+			numRequestsDevice := container.Resources.Requests[corev1.ResourceName(deviceName)]
+			numRequestsCPU += int(numRequestsDevice.Value())
+
+			numLimitsDevice := container.Resources.Limits[corev1.ResourceName(deviceName)]
+			numLimitsCPU += int(numLimitsDevice.Value())
+		}
+
+		if numRequestsCPU == 0 {
+			numResquestsDevice := container.Resources.Requests[CPUResource]
+			numRequestsCPU = int(numResquestsDevice.Value())
+
+			numLimitsDevice := container.Resources.Limits[CPUResource]
+			numLimitsCPU = int(numLimitsDevice.Value())
+		}
+
+		if numRequestsCPU != int(numRequestsPowerProfile.Value()) ||
+			numLimitsCPU != int(numLimitsPowerProfile.Value()) {
 			return "", resourceRequestsMismatchError
 		}
 	}
@@ -398,18 +435,18 @@ func getContainerProfileFromRequests(container corev1.Container, logger *logr.Lo
 	return profileName, nil
 }
 
-func getContainersRequestingExclusiveCPUs(pod *corev1.Pod, logger *logr.Logger) []corev1.Container {
+func getAdmissibleContainers(pod *corev1.Pod, logger *logr.Logger, CustomDevices []string) []corev1.Container {
 
-	logger.V(5).Info("receiving containers requesting exclusive CPUs")
-	containersRequestingExclusiveCPUs := make([]corev1.Container, 0)
+	logger.V(5).Info("Receiving Containers requesting Exclusive CPUs or Custom Devices")
+	admissibleContainers := make([]corev1.Container, 0)
 	containerList := append(pod.Spec.InitContainers, pod.Spec.Containers...)
 	for _, container := range containerList {
-		if doesContainerRequireExclusiveCPUs(pod, &container) {
-			containersRequestingExclusiveCPUs = append(containersRequestingExclusiveCPUs, container)
+		if doesContainerRequireExclusiveCPUs(pod, &container) || validateCustomDevices(pod, &container, CustomDevices) {
+			admissibleContainers = append(admissibleContainers, container)
 		}
 	}
-	logger.V(5).Info("the containers requesting exclusive CPUs are: ", containersRequestingExclusiveCPUs)
-	return containersRequestingExclusiveCPUs
+	logger.V(5).Info("Containers requesting Exclusive CPUs or Curstom Devices are: ", "Containers", admissibleContainers)
+	return admissibleContainers
 
 }
 
@@ -422,6 +459,17 @@ func doesContainerRequireExclusiveCPUs(pod *corev1.Pod, container *corev1.Contai
 	return cpuQuantity.Value()*1000 == cpuQuantity.MilliValue()
 }
 
+func validateCustomDevices(pod *corev1.Pod, container *corev1.Container, CustomDevices []string) bool {
+	presence := false
+	for _, devicePlugin := range CustomDevices {
+		numResources := container.Resources.Requests[corev1.ResourceName(devicePlugin)]
+		if numResources.Value() > 0 {
+			presence = numResources.Value()*1000 == numResources.MilliValue()
+		}
+	}
+	return presence
+}
+
 func getContainerID(pod *corev1.Pod, containerName string) string {
 	for _, containerStatus := range append(pod.Status.InitContainerStatuses, pod.Status.ContainerStatuses...) {
 		if containerStatus.Name == containerName {
-- 
2.34.1

